Compfiles/Imo1970P4.lean

```diff
@@ -21,15 +20,831 @@ equals the product of the nubmers in the other set.
+  by_contra H
+  simp at H
+  obtain ⟨a, b, c⟩ := H
+  have := Nat.Prime.five_le_of_ne_two_of_ne_three pp
+  omega
+
+lemma two_three_five_and_more_is_enough (x : ℕ) (two_does_not_divide : ¬ 2 ∣ x) (three_does_not_divide : ¬ 3 ∣ x) (five_does_not_divide : ¬ 5 ∣ x)
+  (p_gt_5_not_dvd : ∀ (p : ℕ), p.Prime → p > 5 → ¬p ∣ x):
+  ¬ ∃ (p : ℕ), (p.Prime ∧ p ∣ x) := by
+  have p_le_5_not_dvd : ∀ (p : ℕ), p.Prime → p ≤ 5 → ¬ p ∣ x := by
+    intro p
+    intro pp
+    intro p_le_5
+    intro p_div_x
+
+    have p_subset : p ∈ ({2, 3, 5} : Finset ℕ) := enumerate_primes_le_5 p pp p_le_5
+
+    have : p = 2 ∨ p = 3 ∨ p = 5 := by
+      simp_all only [Nat.two_dvd_ne_zero, gt_iff_lt, Finset.mem_insert, Finset.mem_singleton]
+
+    cases this
+    case inl h =>
+      simp_all only [Nat.two_dvd_ne_zero, gt_iff_lt, Finset.mem_insert, Finset.mem_singleton, true_or]
+      omega
+    case inr h =>
+      cases h
+      case inl h =>
+        simp_all only [Nat.two_dvd_ne_zero, gt_iff_lt]
+      case inr h =>
+        simp_all only [Nat.two_dvd_ne_zero, gt_iff_lt, Finset.mem_insert, Finset.mem_singleton]
+  rintro ⟨p, ⟨pp, div⟩⟩
+  have p_gt_5_implies := p_gt_5_not_dvd p pp
+  have p_le_5_implies := p_le_5_not_dvd p pp
+  omega
+
+lemma subsets_must_overlap_pigeonhole (s s1 s2 : Finset ℕ) (predicate_s1: ℕ → Prop) (predicate_s2 : ℕ → Prop)
+                                      [DecidablePred predicate_s1] [DecidablePred predicate_s2]
+                                      (s1_filter : s1 = (s.filter (λ x => predicate_s1 x)))
+                                      (s2_filter : s2 = (s.filter (λ x => predicate_s2 x)))
+                                      (a b c : ℕ) (total_size_exceeds: a + b > c)
+                                      (card_s : s.card = c) (card_s1 : s1.card = a) (card_s2 : s2.card = b):
+                                      ∃ x ∈ s, predicate_s1 x ∧ predicate_s2 x := by
+  have s1_is_subset: s1 ⊆ s := by
+    rw[s1_filter]
+    apply Finset.filter_subset
+  have s2_is_subset: s2 ⊆ s := by
+    rw[s2_filter]
+    apply Finset.filter_subset
+  have lift_x_in_s1 : ∀ x ∈ s, x ∈ s1 ↔ predicate_s1 x := by
+    rw[s1_filter]
+    intro x _
+    subst card_s
+    simp_all only [Finset.mem_filter, true_and]
+  have lift_x_in_s2 : ∀ x ∈ s, x ∈ s2 ↔ predicate_s2 x := by
+    rw[s2_filter]
+    intro x _
+    subst card_s
+    simp_all only [Finset.mem_filter, true_and]
+  apply not_forall_not.mp
+  intro h
+  simp at h
+  have step_1 : (∀ x ∈ s, predicate_s1 x → ¬predicate_s2 x) → (∀ x ∈ s, x ∈ s1 → ¬ x ∈ s2) := by
+    intro left
+    intro x x_in_s
+    intro predicate_s1_fulfills
+    intro predicate_s2_fulfills
+    have part_1 := ((lift_x_in_s1 x x_in_s).mp predicate_s1_fulfills)
+    have part_2 := ((lift_x_in_s2 x x_in_s).mp predicate_s2_fulfills)
+    exact left x x_in_s part_1 part_2
+  have step_2 : (∀ x ∈ s, x ∈ s1 → x ∉ s2) → Disjoint s1 s2 := by
+    intro left
+    dsimp[Disjoint]
+    dsimp[Finset.instHasSubset]
+    simp
```

Comment:
For future-proofing, non-terminal `simp` invocations should be turned into `simp only`. You can do this by changing into `simp?` and then accepting the suggestion in the infoview.

Reply:
I've fixed this in this instance, and all the other instances.